File: docs/snippets/all/tutorials/dna.py
"""The DNA-abacus example from the Log and Ingest tutorial."""

# region: imports
from math import tau

import numpy as np

import rerun as rr
from rerun.utilities import bounce_lerp, build_color_spiral

# endregion: imports


def main() -> None:
    # region: init
    rr.init("rerun_example_dna_abacus", spawn=True)
    # endregion: init

    # The fix for the latest-at lesson — see "Latest-at semantics" in the tutorial.
    # region: latest_at_fix
    rr.set_time("stable_time", duration=0)
    # endregion: latest_at_fix

    NUM_POINTS = 100

    # region: first_points
    points1, colors1 = build_color_spiral(NUM_POINTS)
    points2, colors2 = build_color_spiral(NUM_POINTS, angular_offset=tau * 0.5)

    rr.log("dna/structure/left", rr.Points3D(points1, colors=colors1, radii=0.08))
    rr.log("dna/structure/right", rr.Points3D(points2, colors=colors2, radii=0.08))
    # endregion: first_points

    # region: scaffolding
    rr.log(
        "dna/structure/scaffolding",
        rr.LineStrips3D(np.stack((points1, points2), axis=1), colors=[128, 128, 128]),
    )
    # endregion: scaffolding

    # region: beads
    offsets = np.random.rand(NUM_POINTS)
    beads = [bounce_lerp(points1[n], points2[n], offsets[n]) for n in range(NUM_POINTS)]
    colors = [[int(bounce_lerp(80, 230, offsets[n] * 2))] for n in range(NUM_POINTS)]
    rr.log(
        "dna/structure/scaffolding/beads",
        rr.Points3D(beads, radii=0.06, colors=np.repeat(colors, 3, axis=-1)),
    )
    # endregion: beads

    time_offsets = np.random.rand(NUM_POINTS)

    # region: time_loop
    for i in range(400):
        time = i * 0.01
        rr.set_time("stable_time", duration=time)

        times = np.repeat(time, NUM_POINTS) + time_offsets
        beads = [bounce_lerp(points1[n], points2[n], times[n]) for n in range(NUM_POINTS)]
        colors = [[int(bounce_lerp(80, 230, times[n] * 2))] for n in range(NUM_POINTS)]
        rr.log(
            "dna/structure/scaffolding/beads",
            rr.Points3D(beads, radii=0.06, colors=np.repeat(colors, 3, axis=-1)),
        )
    # endregion: time_loop

    # region: transform_loop
    for i in range(400):
        time = i * 0.01
        rr.set_time("stable_time", duration=time)
        rr.log(
            "dna/structure",
            rr.Transform3D(rotation=rr.RotationAxisAngle(axis=[0, 0, 1], radians=time / 4.0 * tau)),
        )
    # endregion: transform_loop


if __name__ == "__main__":
    main()

File: docs/snippets/all/tutorials/dna.rs
//! The DNA-abacus example from the Log and Ingest tutorial.

// region: imports
use std::f32::consts::TAU;

use itertools::Itertools as _;
use rand::Rng as _;
use rerun::{
    demo_util::{bounce_lerp, color_spiral},
    external::glam,
};
// endregion: imports

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // region: init
    let rec = rerun::RecordingStreamBuilder::new("rerun_example_dna_abacus").spawn()?;
    // endregion: init

    // The fix for the latest-at lesson — see "Latest-at semantics" in the tutorial.
    // region: latest_at_fix
    rec.set_duration_secs("stable_time", 0.0);
    // endregion: latest_at_fix

    const NUM_POINTS: usize = 100;

    // region: first_points
    let (points1, colors1) = color_spiral(NUM_POINTS, 2.0, 0.02, 0.0, 0.1);
    let (points2, colors2) = color_spiral(NUM_POINTS, 2.0, 0.02, TAU * 0.5, 0.1);

    rec.log(
        "dna/structure/left",
        &rerun::Points3D::new(points1.iter().copied())
            .with_colors(colors1)
            .with_radii([0.08]),
    )?;
    rec.log(
        "dna/structure/right",
        &rerun::Points3D::new(points2.iter().copied())
            .with_colors(colors2)
            .with_radii([0.08]),
    )?;
    // endregion: first_points

    // region: scaffolding
    let lines: Vec<[glam::Vec3; 2]> = points1
        .iter()
        .zip(&points2)
        .map(|(&p1, &p2)| (p1, p2).into())
        .collect_vec();

    rec.log(
        "dna/structure/scaffolding",
        &rerun::LineStrips3D::new(lines.iter().copied())
            .with_colors([rerun::Color::from_rgb(128, 128, 128)]),
    )?;
    // endregion: scaffolding

    // region: beads
    let mut rng = rand::rng();
    let offsets = (0..NUM_POINTS).map(|_| rng.random::<f32>()).collect_vec();

    let beads = lines
        .iter()
        .zip(&offsets)
        .map(|(&[p1, p2], &offset)| bounce_lerp(p1, p2, offset))
        .collect_vec();
    let colors = offsets
        .iter()
        .map(|&offset| bounce_lerp(80.0, 230.0, offset * 2.0) as u8)
        .map(|c| rerun::Color::from_rgb(c, c, c))
        .collect_vec();

    rec.log(
        "dna/structure/scaffolding/beads",
        &rerun::Points3D::new(beads)
            .with_colors(colors)
            .with_radii([0.06]),
    )?;
    // endregion: beads

    // region: time_loop
    for i in 0..400 {
        let time = i as f32 * 0.01;

        rec.set_duration_secs("stable_time", time);

        let times = offsets.iter().map(|offset| time + offset).collect_vec();
        let beads = lines
            .iter()
            .zip(&times)
            .map(|(&[p1, p2], &time)| bounce_lerp(p1, p2, time))
            .collect_vec();
        let colors = times
            .iter()
            .map(|time| bounce_lerp(80.0, 230.0, time * 2.0) as u8)
            .map(|c| rerun::Color::from_rgb(c, c, c))
            .collect_vec();

        rec.log(
            "dna/structure/scaffolding/beads",
            &rerun::Points3D::new(beads)
                .with_colors(colors)
                .with_radii([0.06]),
        )?;
    }
    // endregion: time_loop

    // region: transform_loop
    for i in 0..400 {
        let time = i as f32 * 0.01;

        rec.set_duration_secs("stable_time", time);

        rec.log(
            "dna/structure",
            &rerun::archetypes::Transform3D::from_rotation(rerun::RotationAxisAngle::new(
                glam::Vec3::Z,
                rerun::Angle::from_radians(time / 4.0 * TAU),
            )),
        )?;
    }
    // endregion: transform_loop

    Ok(())
}
